File: core/tab_defaults.py
"""Settings '기본값' 패널의 저장소(config/tab_defaults.json) — Qt 없는 순수 로직.

누가 무엇을 읽나(모두 실제로 쓰인다 — audit #140):
- T2I steps/cfg/width/height/seed/sampler/scheduler: 시작 시 빈 위젯(첫 실행이면 전부)에 적용
  (generator_main._apply_tab_defaults_to_empty_widgets).
- denoising: I2I 화면의 Denoising 초기값(I2IView).
- brushSize/effectStrength/yoloConf/snapRadius: 에디터 마스크 브러시·효과 세기·YOLO 신뢰도·
  자석 올가미 스냅 반경 초기값(EditorView/EffectPanel/EditorCanvas).
- hires_enabled/ad_enabled/sam3_enabled(+레거시 ad_s1/ad_s2): 첫 실행(prompt_settings.json 없음)에만.

저장은 **부분 병합**이다. Settings(keep-alive)가 들고 있던 옛 값 전체를 1.5초 뒤 통째로 써서
'전역 저장'이 방금 T2I 값으로 갱신한 steps/cfg/… 를 되덮던 양방향 덮어쓰기를 막는다 — 화면은
바뀐 키만 보내고, 여기서 기존 파일 위에 병합한다.

의미가 없어진 키(defaultRating: 검색 등급 필터는 ui_prefs.ratingFilter 가 주인 /
negpip_enabled: NegPiP 는 상시 적용)는 다음 저장 때 파일에서 지운다.
"""

from __future__ import annotations

import math
import os
from collections.abc import Mapping
from typing import Any, Callable

from utils.atomic_json import atomic_write_json, load_json_safe


def _int(lo: int, hi: int) -> Callable[[Any], int]:
    def conv(value: Any) -> int:
        number = float(value)
        if not math.isfinite(number):
            raise ValueError('not finite')
        return max(lo, min(hi, int(round(number))))
    return conv


def _float(lo: float, hi: float) -> Callable[[Any], float]:
    def conv(value: Any) -> float:
        number = float(value)
        if not math.isfinite(number):
            raise ValueError('not finite')
        return max(lo, min(hi, round(number, 4)))
    return conv


def _text(value: Any) -> str:
    return str(value if value is not None else '').strip()


def _seed(value: Any) -> str:
    text = _text(value)
    return text or '-1'


def _bool(value: Any) -> bool:
    if isinstance(value, str):
        return value.strip().lower() in {'1', 'true', 'yes', 'on'}
    return bool(value)


#: 키 → 정규화 함수. 프론트 frontend/src/utils/tabDefaults.ts 의 FACTORY_DEFAULTS 와 같은 키.
TAB_DEFAULT_FIELDS: dict[str, Callable[[Any], Any]] = {
    'steps': _int(1, 500),
    'cfg': _float(0.0, 100.0),
    'width': _int(64, 8192),
    'height': _int(64, 8192),
    'seed': _seed,
    'sampler': _text,
    'scheduler': _text,
    'denoising': _float(0.0, 1.0),
    'brushSize': _int(1, 500),
    'effectStrength': _int(1, 100),
    'yoloConf': _float(0.01, 1.0),
    'snapRadius': _int(1, 100),
    'hires_enabled': _bool,
    'ad_enabled': _bool,
    'sam3_enabled': _bool,
    # 화면에는 없지만 첫 실행 적용이 읽는 레거시 키 — 보존한다.
    'ad_s1_enabled': _bool,
    'ad_s2_enabled': _bool,
}

#: 더는 의미가 없어 저장 때 지우는 키.
REMOVED_TAB_DEFAULT_KEYS = frozenset({'defaultRating', 'negpip_enabled'})

#: '전역 저장'이 현재 T2I 값으로 갱신하는 키(f90af4319 'Default 연동').
T2I_LINKED_KEYS = ('steps', 'cfg', 'width', 'height', 'seed')


def default_tab_defaults_path() -> str:
    return os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'config', 'tab_defaults.json')


def normalize_tab_defaults(raw: Any, *, keep_unknown: bool = False) -> dict[str, Any]:
    """알려진 키만 정규화해 돌려준다. 값이 깨졌으면 그 키를 뺀다(기본값은 호출자가 정한다)."""
    if not isinstance(raw, Mapping):
        return {}
    out: dict[str, Any] = {}
    for key, value in raw.items():
        name = str(key)
        if name in REMOVED_TAB_DEFAULT_KEYS:
            continue
        conv = TAB_DEFAULT_FIELDS.get(name)
        if conv is None:
            if keep_unknown:
                out[name] = value
            continue
        try:
            out[name] = conv(value)
        except (TypeError, ValueError, OverflowError):
            continue
    return out


def load_tab_defaults(path: str | None = None) -> dict[str, Any]:
    data = load_json_safe(path or default_tab_defaults_path(), {}, backup_corrupt=False)
    return normalize_tab_defaults(data)


def merge_tab_defaults(existing: Any, patch: Any) -> dict[str, Any]:
    """기존 파일 위에 patch 의 알려진 키만 덮는다. 모르는 기존 키는 보존, 폐기 키는 지운다."""
    merged = normalize_tab_defaults(existing, keep_unknown=True)
    merged.update(normalize_tab_defaults(patch))
    return merged


def save_tab_defaults_patch(patch: Any, path: str | None = None) -> dict[str, Any]:
    """patch 를 병합 저장하고 저장된 (정규화된) 값을 돌려준다."""
    target = path or default_tab_defaults_path()
    existing = load_json_safe(target, {}, backup_corrupt=False)
    merged = merge_tab_defaults(existing, patch)
    atomic_write_json(target, merged)
    return normalize_tab_defaults(merged)


def sync_t2i_defaults(values: Mapping[str, Any], path: str | None = None) -> dict[str, Any] | None:
    """'전역 저장' 연동 — 현재 T2I steps/cfg/width/height/seed 를 기본값으로 반영한다.

    파일이 없으면(기본값을 한 번도 저장하지 않음) 만들지 않는다 — 예전 동작과 같다.
    """
    target = path or default_tab_defaults_path()
    if not os.path.exists(target):
        return None
    patch = {key: values[key] for key in T2I_LINKED_KEYS if key in values and values[key] not in (None, '')}
    return save_tab_defaults_patch(patch, target)

File: frontend/src/utils/tabDefaults.ts
/**
 * Settings '기본값' 패널(config/tab_defaults.json)의 프론트 규칙 — 파이썬 core/tab_defaults.py 와 같은 키.
 *
 * 누가 읽나(audit #140 — 예전엔 대부분 아무도 안 읽었다):
 *  - steps/cfg/width/height/seed/sampler/scheduler → 시작 시 T2I 빈 칸(첫 실행이면 전부)
 *  - denoising → I2I Denoising 초기값
 *  - brushSize/effectStrength/yoloConf/snapRadius → 에디터 브러시·효과 세기·YOLO 신뢰도·스냅 반경
 *  - hires/ad/sam3_enabled → 첫 실행(prompt_settings.json 없음)에만
 *
 * 저장은 바뀐 키만 보낸다(diffTabDefaults) — keep-alive Settings 가 들고 있던 옛 값이 '전역 저장'이
 * 갱신한 T2I 기본값을 되덮던 양방향 덮어쓰기를 막는다. 파이썬이 기존 파일 위에 병합한다.
 * 값이 바뀌면 TAB_DEFAULTS_EVENT 로 알린다 — 에디터·I2I 는 사용자가 건드리지 않은 값만 따라간다.
 */

export interface TabDefaults {
  steps: number
  cfg: number
  width: number
  height: number
  seed: string
  sampler: string
  scheduler: string
  denoising: number
  brushSize: number
  effectStrength: number
  yoloConf: number
  snapRadius: number
  hires_enabled: boolean
  ad_enabled: boolean
  sam3_enabled: boolean
}

export const FACTORY_TAB_DEFAULTS: Readonly<TabDefaults> = Object.freeze({
  steps: 20, cfg: 7, width: 1024, height: 1024, seed: '-1',
  sampler: '', scheduler: '', denoising: 0.75,
  brushSize: 20, effectStrength: 15, yoloConf: 0.25, snapRadius: 12,
  hires_enabled: false, ad_enabled: false, sam3_enabled: false,
})

export const TAB_DEFAULT_KEYS = Object.keys(FACTORY_TAB_DEFAULTS) as (keyof TabDefaults)[]

/** 창 이벤트 — detail 은 정규화된 TabDefaults(저장 직후 값). */
export const TAB_DEFAULTS_EVENT = 'tabDefaultsChanged'

type Range = readonly [number, number]
const NUMBER_RANGES: Partial<Record<keyof TabDefaults, { range: Range; int: boolean }>> = {
  steps: { range: [1, 500], int: true },
  cfg: { range: [0, 100], int: false },
  width: { range: [64, 8192], int: true },
  height: { range: [64, 8192], int: true },
  denoising: { range: [0, 1], int: false },
  brushSize: { range: [1, 500], int: true },
  effectStrength: { range: [1, 100], int: true },
  yoloConf: { range: [0.01, 1], int: false },
  snapRadius: { range: [1, 100], int: true },
}
const BOOL_KEYS = new Set<keyof TabDefaults>(['hires_enabled', 'ad_enabled', 'sam3_enabled'])

function normalizeValue(key: keyof TabDefaults, raw: unknown): unknown {
  const spec = NUMBER_RANGES[key]
  if (spec) {
    const n = typeof raw === 'number' ? raw : Number(String(raw ?? '').trim())
    if (!Number.isFinite(n) || (typeof raw === 'string' && !raw.trim())) return undefined
    const clamped = Math.min(spec.range[1], Math.max(spec.range[0], n))
    return spec.int ? Math.round(clamped) : Math.round(clamped * 10000) / 10000
  }
  if (BOOL_KEYS.has(key)) {
    if (typeof raw === 'string') return ['1', 'true', 'yes', 'on'].includes(raw.trim().toLowerCase())
    return !!raw
  }
  if (key === 'seed') {
    const s = String(raw ?? '').trim()
    return s || '-1'
  }
  return String(raw ?? '').trim()
}

/** 파일/브리지 값 → 완전한 TabDefaults(없거나 깨진 키는 공장값). 폐기된 키는 버린다. */
export function normalizeTabDefaults(raw: unknown): TabDefaults {
  const out: TabDefaults = { ...FACTORY_TAB_DEFAULTS }
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return out
  const src = raw as Record<string, unknown>
  for (const key of TAB_DEFAULT_KEYS) {
    if (!(key in src)) continue
    const v = normalizeValue(key, src[key])
    if (v !== undefined) (out as any)[key] = v
  }
  return out
}

/** synced(마지막으로 파일과 맞춘 값)와 다른 키만 — 저장 요청에 실을 부분 값. 바뀐 게 없으면 null. */
export function diffTabDefaults(current: TabDefaults, synced: TabDefaults): Partial<TabDefaults> | null {
  const out: Partial<TabDefaults> = {}
  for (const key of TAB_DEFAULT_KEYS) {
    const a = normalizeValue(key, current[key])
    if (a === undefined) continue   // 입력 중인 빈 숫자 칸 — 저장하지 않는다
    if (a !== normalizeValue(key, synced[key])) (out as any)[key] = a
  }
  return Object.keys(out).length ? out : null
}

/** 에디터가 쓰는 네 값 — YOLO 신뢰도는 EffectPanel 슬라이더 단위(%)로. */
export function editorDefaultsFrom(d: TabDefaults): { brushSize: number; effectStrength: number; detectConf: number; snapRadius: number } {
  return {
    brushSize: d.brushSize,
    effectStrength: d.effectStrength,
    detectConf: Math.min(100, Math.max(1, Math.round(d.yoloConf * 100))),
    snapRadius: d.snapRadius,
  }
}

/**
 * 기본값이 바뀌었을 때 화면 값을 따라가게 할지 — 사용자가 이전 기본값에서 손대지 않았을 때만.
 * (이미 바꾼 값을 설정 화면 저장이 덮어쓰면 안 된다.)
 */
export function followDefault<T>(current: T, previousDefault: T, nextDefault: T): T {
  return current === previousDefault ? nextDefault : current
}

type BackendLike = { getTabDefaults?: (cb: (json: string) => void) => void }

/** 브리지에서 읽기 — 실패하면 공장값. */
export function readTabDefaults(backend: BackendLike | null | undefined): Promise<TabDefaults> {
  return new Promise(resolve => {
    if (!backend || typeof backend.getTabDefaults !== 'function') { resolve({ ...FACTORY_TAB_DEFAULTS }); return }
    try {
      backend.getTabDefaults((json: string) => {
        try { resolve(normalizeTabDefaults(JSON.parse(json || '{}'))) } catch { resolve({ ...FACTORY_TAB_DEFAULTS }) }
      })
    } catch {
      resolve({ ...FACTORY_TAB_DEFAULTS })
    }
  })
}

/** 저장 직후 다른 화면에 알린다. */
export function announceTabDefaults(d: TabDefaults, target: Pick<Window, 'dispatchEvent'> | null =
  (typeof window === 'undefined' ? null : window)): void {
  try { target?.dispatchEvent(new CustomEvent(TAB_DEFAULTS_EVENT, { detail: { ...d } })) } catch {}
}
